File: app/src/main/java/com/oliver/weatherapp/Injector.java
package com.oliver.weatherapp;

import android.content.Context;

import com.oliver.weatherapp.data.local.WeatherDatabase;
import com.oliver.weatherapp.data.remote.WeatherDataSource;
import com.oliver.weatherapp.data.repositories.CitiesRepository;
import com.oliver.weatherapp.data.repositories.WeatherRepository;
import com.oliver.weatherapp.screens.forecast.WeatherViewModelFactory;
import com.oliver.weatherapp.screens.home.CitiesViewModelFactory;

public abstract class Injector {

    public static CitiesRepository provideCitiesRepository(Context context) {
        WeatherDatabase database = WeatherDatabase.getInstance(context);
        AppExecutors executors = AppExecutors.getInstance();
        return CitiesRepository.getInstance(executors, database.citiesDao());
    }

    public static WeatherRepository provideWeatherRepository(Context context) {
        WeatherDatabase database = WeatherDatabase.getInstance(context);
        AppExecutors executors = AppExecutors.getInstance();
        return WeatherRepository.getInstance(executors, database.weatherDao());
    }

    public static WeatherDataSource provideWeatherDataSource(Context context) {
        AppExecutors executors = AppExecutors.getInstance();
        return WeatherDataSource.getInstance(context, executors);
    }

    public static CitiesViewModelFactory provideCitiesViewModelFactory(Context context) {
        CitiesRepository repository = provideCitiesRepository(context.getApplicationContext());
        return new CitiesViewModelFactory(repository);
    }

    public static WeatherViewModelFactory provideWeatherViewModelFactory(Context context, long cityID) {
        WeatherRepository repository = provideWeatherRepository(context.getApplicationContext());
        return new WeatherViewModelFactory(repository, cityID);
    }
}

File: app/src/main/java/com/oliver/weatherapp/Injector.kt
package com.oliver.weatherapp

import android.content.Context

import com.oliver.weatherapp.data.local.WeatherDatabase
import com.oliver.weatherapp.data.local.model.CityEntry
import com.oliver.weatherapp.data.remote.WeatherDataSource
import com.oliver.weatherapp.data.repositories.CitiesRepository
import com.oliver.weatherapp.data.repositories.WeatherRepository
import com.oliver.weatherapp.screens.forecast.WeatherViewModelFactory
import com.oliver.weatherapp.screens.home.CitiesViewModelFactory

object Injector {

    private fun provideCitiesRepository(context: Context): CitiesRepository {
        val database = WeatherDatabase.getInstance(context)
        val executors = AppExecutors.getInstance()
        return CitiesRepository.getInstance(executors, database.citiesDao())
    }

    private fun provideWeatherRepository(context: Context): WeatherRepository {
        val database = WeatherDatabase.getInstance(context)
        val executors = AppExecutors.getInstance()
        val dataSource = provideWeatherDataSource()
        return WeatherRepository.getInstance(executors, database.weatherDao(), dataSource)
    }

    private fun provideWeatherDataSource(): WeatherDataSource {
        val executors = AppExecutors.getInstance()
        return WeatherDataSource.getInstance(executors)
    }

    fun provideCitiesViewModelFactory(context: Context): CitiesViewModelFactory {
        val repository = provideCitiesRepository(context.applicationContext)
        return CitiesViewModelFactory(repository)
    }

    fun provideWeatherViewModelFactory(context: Context, city: CityEntry): WeatherViewModelFactory {
        val repository = provideWeatherRepository(context.applicationContext)
        return WeatherViewModelFactory(repository, city)
    }
}
